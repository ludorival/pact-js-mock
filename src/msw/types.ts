/* eslint-disable @typescript-eslint/no-explicit-any */
import type { GraphQLError } from 'graphql'

export type GraphQLPayload = Record<string, any>
export type GraphQLResponse<T = any> = {
  data?: T
  errors?: GraphQLError[]
}
