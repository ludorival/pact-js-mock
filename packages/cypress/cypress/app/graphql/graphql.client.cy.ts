/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { createTodo, fetchTodos, todoById } from './graphql.client'
import {
  createTodoWillSucceed,
  emptyTodos,
  multipleTodos,
  pact,
  todoByIdFound,
  todoByIdNotFound,
  todosWillRaiseTechnicalFailure,
} from './handlers'

before(() => {
  pact.reset()
})

after(() => {
  const pactFile = pact.generatePactFile()
  cy.writeFile(`pacts/${pact.name}.json`, pactFile)
  cy.fixture('test-consumer-graphql-provider.json').then((expectedPact) => {
    expect(pactFile).to.deep.equal(expectedPact)
  })
})
describe('To-Do list GraphQL API client', () => {
  describe('fetchTodos', () => {
    it('should fetch all To-Do items', () => {
      // use multipleTodos handlers from contracts
      cy.intercept('GET', '*/todos', multipleTodos).as('multipleTodos')

      // call the fetchTodos function and get the actual data
      cy.request(fetchTodos())

      // expect the actual data to match the expected data
      cy.wait('@multipleTodos')
        .its('response')
        .its('statusCode')
        .should('be.equal', 200)
    })

    it('should get a technical failure the first time and an empty todo list', () => {
      // use todosWillRaiseTechnicalFailure and emptyTodos handlers from contracts
      cy.intercept('GET', '*/todos', todosWillRaiseTechnicalFailure).as(
        'todosWillRaiseTechnicalFailure'
      )
      // call first time fetchTodos should return an error

      cy.request(fetchTodos())

      cy.wait('@todosWillRaiseTechnicalFailure')
        .its('response')
        .its('statusMessage')
        .should('be.equal', 'Request failed with status code 500')
      cy.intercept('GET', '*/todos', emptyTodos).as('emptyTodos')

      // call the fetchTodos function and get the actual data
      cy.request(fetchTodos())

      // expect the actual data to match the expected data
      cy.wait('@emptyTodos').its('response.body').should('be.equal', [])
    })
  })

  describe('createTodo', () => {
    it('should create a new To-Do item', () => {
      // use createTodoWillSucceed handlers from contracts
      cy.intercept('POST', '*/todos', createTodoWillSucceed).as(
        'createTodoWillSucceed'
      )

      // call the createTodo function and get the actual data
      cy.request(createTodo('Buy groceries'))

      // expect the actual data to match the status code
      cy.wait('@createTodoWillSucceed')
        .its('response')
        .its('statusCode')
        .should('be.equal', 201)
    })
  })

  describe('todoById', () => {
    it('should get a todo by its id', () => {
      // use todoByIdFound handlers from contracts
      cy.intercept('GET', '*/todos/*', todoByIdFound).as('todoByIdFound')

      // call the todoById function and get the actual data
      cy.request(todoById('1'))

      // expect the actual data to match the expected status code
      cy.wait('@todoByIdFound')
        .its('response')
        .its('statusCode')
        .should('be.equal', 200)
    })

    it('should get an error when getting a todo does not found it', () => {
      // use todoByIdFound handlers from contracts
      cy.intercept('GET', '*/todos/*', todoByIdNotFound).as('todoByIdNotFound')

      // call the todoById function and get the actual data
      cy.request(todoById('1'))
      // expect the actual data to match the expected status code
      cy.wait('@todoByIdNotFound')
        .its('response')
        .its('statusMessage')
        .should('be.equal', 'Request failed with status code 404')
    })
  })
})
