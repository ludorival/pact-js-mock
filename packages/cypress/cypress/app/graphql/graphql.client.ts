const url = 'https://pact.msw.example.com/graphql'

function graphql<TVariables = unknown>({
  query,
  variables,
}: {
  query: string
  variables?: TVariables
}): Partial<Cypress.RequestOptions> {
  return {
    method: 'POST',
    url,
    body: { query, variables },
  }
}
// define a function to fetch all To-Do items
export function fetchTodos() {
  // define the GraphQL query
  const query = `
    query todos {
      todos {
        id
        title
        description
        completed
      }
    }
  `

  // execute the query and get the data
  return graphql({ query })
}

// define a function to fetch all To-Do items
export function todoById(id: string) {
  // define the GraphQL query
  const query = `
    query todoById($id: ID) {
      todoById(id: $id) {
        id
        title
        description
        completed
      }
    }
  `

  // execute the query and get the data
  return graphql<{ id: string }>({
    query,
    variables: { id },
  })
}
// define a function to create a new To-Do item
export function createTodo(title: string, description?: string) {
  // define the GraphQL mutation
  const query = `
    mutation createTodo($title: String!, $description: String) {
      createTodo(input: { title: $title, description: $description }) {
        id
        title
        description
        completed
      }
    }
  `

  // execute the mutation and get the data
  return graphql<{ title: string; description?: string }>({
    query,
    variables: { title, description },
  })
}
