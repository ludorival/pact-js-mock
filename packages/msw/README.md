<h1 align="center">Welcome to pact-mock-js.msw üëã</h1>
<p>
  <img alt="Version" src="https://img.shields.io/github/v/release/ludorival/pact-mock-js/msw" />
  <a href="https://github.com/ludorival/pact-mock-js/msw#readme" target="_blank">
    <img alt="Documentation" src="https://img.shields.io/badge/documentation-yes-brightgreen.svg" />
  </a>
  <a href="https://github.com/ludorival/pact-mock-js/msw/graphs/commit-activity" target="_blank">
    <img alt="Maintenance" src="https://img.shields.io/badge/Maintained%3F-yes-green.svg" />
  </a>
  <a href="https://github.com/ludorival/pact-mock-js/msw/blob/master/LICENSE" target="_blank">
    <img alt="License: BSD--3--Clause" src="https://img.shields.io/github/license/ludorival/pact-mock-js/msw" />
  </a>
</p>

# pact-mock-js.msw

`pact-mock-js.msw` is a Node.js library that allows you to build [Pact](https://docs.pact.io/) contracts by leveraging your existings [msw](https://mswjs.io/) mocks. This library provides an easy way to generate contracts that can be used for testing and verifying API interactions between consumer and provider.

## Install pact-mock-js.msw

```sh
yarn add -D pact-mock-js.msw
```

## Usage

Here is an example of how to use pact-mock-js.msw:

```js
import { setupServer, rest } from 'msw/node'
import { Pact } from 'pact-mock-js.msw'
import { writeFile } from 'fs'

const server = setupServer()

const pact = new Pact({
  consumer: { name: 'test-consumer' },
  provider: { name: 'rest-provider' },
  metadata: { pactSpecification: { version: '2.0.0' } },
})

beforeAll(() => {
  server.listen()
  pact.reset()
})

afterEach(() => {
  server.resetHandlers()
})

afterAll(() => {
  server.close()
  // Write the pact file wherever you want
  fs.writeFile(
    `pacts/${pact.name}.json`,
    JSON.stringify(pact.generatePactFile())
  )
})

it('get all movies', async () => {
  const mockMovies = rest.get(
    '*/movies',
    pact.toResolver({
      description: 'a request to list all movies',
      response: [
        {
          id: 1,
          name: 'Movie 1',
          year: 2008,
        },
        {
          id: 2,
          name: 'Movie 2',
          year: 2008,
        },
      ],
    })
  )

  server.use(mockMovies)

  const movies = await fetchMovies()

  expect(movies).toEqual([
    {
      id: 1,
      name: 'Movie 1',
      year: 2008,
    },
    {
      id: 2,
      name: 'Movie 2',
      year: 2008,
    },
  ])
})
```

You can find more example to mock

- [A Rest API](./packages/msw/rest/rest.client.test.ts)
- [A GraphQL API](./packages/msw/graphql/graphql.client.test.ts)

## Author

üë§ **Ludovic Dorival**

- Github: [@ludorival](https://github.com/ludorival)

## Show your support

Give a ‚≠êÔ∏è if this project helped you!

## üìù License

Copyright ¬© 2021 [Ludovic Dorival](https://github.com/ludorival).<br />
This project is [BSD--3--Clause](https://github.com/ludorival/pact-mock-js/msw/blob/master/LICENSE) licensed.

---

_This README was generated with ‚ù§Ô∏è by [readme-md-generator](https://github.com/kefranabg/readme-md-generator)_
