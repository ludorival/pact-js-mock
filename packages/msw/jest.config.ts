import config from '../../jest.config.base'
/*
 * For a detailed explanation regarding each configuration property and type check, visit:
 * https://jestjs.io/docs/en/configuration.html
 */

export default {
  ...config,
  roots: ['<rootDir>/src/'],
}
