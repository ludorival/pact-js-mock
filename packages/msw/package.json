{
  "name": "@pact-mock-js/msw",
  "version": "0.0.1",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "test": "jest --testPathPattern='packages/msw'"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "description": "",
  "dependencies": {
    "@pact-mock-js/core": "0.0.1",
    "lodash": "^4.17.21"
  },
  "devDependencies": {
    "@types/lodash": "^4.14.192",
    "axios": "^1.7.7",
    "graphql": "^16.6.0",
    "msw": "^1.2.1"
  }
}
