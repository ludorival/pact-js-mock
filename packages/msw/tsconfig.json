{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "dist",
    "rootDir": "src",
    "baseUrl": "src",
    "paths": {
      "pact-msw": ["index.ts"]
    },
    /* List of root folders whose combined content represents the structure of the project at runtime. */
    "types": ["node", "jest"]
  },
  "include": ["src"]
}
