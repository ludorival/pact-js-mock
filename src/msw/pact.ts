import { omit } from 'lodash'
import {
  DefaultBodyType,
  HttpResponse,
  HttpResponseInit,
  StrictRequest,
  StrictResponse,
} from 'msw'
import { Pact as BasePact, buildResponse } from '../core'
import {
  InputPact,
  InteractionFor,
  MinimalInteraction,
  Options,
  PactFile,
  PactV2,
  PactV3,
  PactV4,
  Request,
  ToRecordInteraction,
} from '../types'
type Info<
  R extends DefaultBodyType = DefaultBodyType,
  Extra extends Record<string, unknown> = Record<string, unknown>,
> = {
  request: StrictRequest<R>
} & Extra
export class Pact<P extends PactFile> extends BasePact<P> {
  constructor(pact: InputPact<P>, options?: Options) {
    super(pact, options)
  }

  toResolver<T extends object, R extends DefaultBodyType = DefaultBodyType>(
    input: MinimalInteraction<InteractionFor<P, T>> | T,
  ): (info: Info<R>) => Promise<StrictResponse<T>> {
    const version = this.version
    return async (info) => {
      const interaction =
        'response' in input ? input : buildResponse(input, version)
      const response = await this.toResponse(
        interaction as MinimalInteraction<InteractionFor<P, T>>,
        info,
      )
      return response
    }
  }

  async toResponse<
    TResponse extends DefaultBodyType,
    TRequest extends DefaultBodyType = DefaultBodyType,
  >(
    interaction: MinimalInteraction<InteractionFor<P, TResponse>>,
    info: Info<TRequest>,
    initOptions?: HttpResponseInit,
  ): Promise<StrictResponse<TResponse>> {
    const toRecord = {
      ...interaction,
      request: await toRequest(info),
    } as ToRecordInteraction<InteractionFor<P>>
    this.record(toRecord)

    const responseV4 = (interaction.response as { body?: { content: unknown } })
      .body?.content
      ? (interaction.response as PactV4.ResponseClass)
      : null
    const response = responseV4
      ? null
      : (interaction.response as PactV2.Response | PactV3.Response)
    const content = responseV4?.body?.content || response?.body

    const responseStatus = responseV4?.status || response?.status || 200

    return HttpResponse.json(content, {
      headers: response?.headers || responseV4?.headers,
      status: responseStatus,
      ...initOptions,
    })
  }
}

async function toRequest<R extends DefaultBodyType = DefaultBodyType>(
  info: Info<R>,
): Promise<Request> {
  const url = new URL(info.request.url)
  const path = url.pathname
  const query = url.searchParams.toString() || undefined
  const body = info.query
    ? omit(info, 'request', 'requestId')
    : await info.request.json().catch(() => info.request.body)
  const headers: Record<string, unknown> = {}
  info.request.headers.forEach((value, key) => (headers[key] = value))
  return {
    method: info.request.method,
    path,
    headers,
    body,
    query,
  } as Request
}
