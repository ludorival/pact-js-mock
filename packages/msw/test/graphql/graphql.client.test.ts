/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { setupServer } from 'msw/node'
import { createTodo, fetchTodos, todoById } from './graphql.client'
import {
  createTodoWillSucceed,
  emptyTodos,
  multipleTodos,
  pact,
  todoByIdFound,
  todoByIdNotFound,
  todosWillRaiseTechnicalFailure,
} from './handlers'

const server = setupServer()

beforeAll(async () => {
  await pact.reset()
  server.listen()
})

afterEach(() => {
  server.resetHandlers()
})

afterAll(async () => {
  server.close()
  const pactFile = await pact.generatePactFile()
  expect(pactFile).toMatchSnapshot()
})

describe('To-Do list GraphQL API client', () => {
  describe('fetchTodos', () => {
    it('should fetch all To-Do items', async () => {
      // use multipleTodos handlers from contracts
      server.use(multipleTodos)

      // call the fetchTodos function and get the actual data
      const actualData = await fetchTodos()

      // expect the actual data to match the expected data
      expect(actualData).toMatchSnapshot()
    })

    it('should get a technical failure the first time and an empty todo list', async () => {
      // use todosWillRaiseTechnicalFailure and emptyTodos handlers from contracts
      server.use(todosWillRaiseTechnicalFailure, emptyTodos)

      // call first time fetchTodos should return an error
      expect.assertions(1)
      fetchTodos().catch((e) =>
        expect(e).toMatchObject({
          message: 'Request failed with status code 500',
        })
      )
      // call the fetchTodos function and get the actual data
      const actualData = await fetchTodos()

      // expect the actual data to match the expected data
      expect(actualData).toEqual([])
    })
  })

  describe('createTodo', () => {
    it('should create a new To-Do item', async () => {
      // use createTodoWillSucceed handlers from contracts
      server.use(createTodoWillSucceed)

      // call the createTodo function and get the actual data
      const actualData = await createTodo('Buy groceries')

      // expect the actual data to match the expected data
      expect(actualData).toMatchSnapshot()
    })
  })

  describe('todoById', () => {
    it('should get a todo by its id', async () => {
      // use todoByIdFound handlers from contracts
      server.use(todoByIdFound)

      // call the todoById function and get the actual data
      const actualData = await todoById('1')

      // expect the actual data to match the expected data
      expect(actualData).toMatchSnapshot()
    })

    it('should get an error when getting a todo does not found it', async () => {
      // use todoByIdFound handlers from contracts
      server.use(todoByIdNotFound)

      // call the todoById function and get the actual data
      try {
        await todoById('1')
        fail('Should never reach')
      } catch (e: unknown) {
        expect(e).toMatchObject({ message: 'The todo item 1 is not found' })
      }
    })
  })
})
