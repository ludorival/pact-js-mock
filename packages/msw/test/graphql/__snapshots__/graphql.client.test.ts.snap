// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`To-Do list GraphQL API client createTodo should create a new To-Do item 1`] = `
{
  "completed": false,
  "description": "Milk, bread, eggs, cheese",
  "id": "1",
  "title": "Buy groceries",
}
`;

exports[`To-Do list GraphQL API client fetchTodos should fetch all To-Do items 1`] = `
[
  {
    "completed": false,
    "description": "Milk, bread, eggs, cheese",
    "id": "1",
    "title": "Buy groceries",
  },
  {
    "completed": true,
    "description": "",
    "id": "2",
    "title": "Do laundry",
  },
  {
    "completed": false,
    "description": "Fix leaky faucet in the bathroom",
    "id": "3",
    "title": "Call plumber",
  },
]
`;

exports[`To-Do list GraphQL API client todoById should get a todo by its id 1`] = `
{
  "completed": false,
  "description": "Milk, bread, eggs, cheese",
  "id": "1",
  "title": "Buy groceries",
}
`;

exports[`To-Do list GraphQL API client todoById should get an error when getting a todo does not found it 1`] = `
{
  "consumer": {
    "name": "test-consumer",
  },
  "interactions": [
    {
      "description": "empty todo list",
      "request": {
        "body": {
          "query": "
    query todos {
      todos {
        id
        title
        description
        completed
      }
    }
  ",
        },
        "headers": {
          "content-type": "application/json",
        },
        "method": "POST",
        "path": "/graphql",
      },
      "response": {
        "body": {
          "data": {
            "todos": [],
          },
        },
        "status": 200,
      },
    },
    {
      "description": "graphql api returns a 500 http error",
      "providerState": "will return a 500 http error",
      "request": {
        "body": {
          "query": "
    query todos {
      todos {
        id
        title
        description
        completed
      }
    }
  ",
        },
        "headers": {
          "content-type": "application/json",
        },
        "method": "POST",
        "path": "/graphql",
      },
      "response": {
        "status": 500,
      },
    },
    {
      "description": "multiple todo list",
      "request": {
        "body": {
          "query": "
    query todos {
      todos {
        id
        title
        description
        completed
      }
    }
  ",
        },
        "headers": {
          "content-type": "application/json",
        },
        "method": "POST",
        "path": "/graphql",
      },
      "response": {
        "body": {
          "data": {
            "todos": [
              {
                "completed": false,
                "description": "Milk, bread, eggs, cheese",
                "id": "1",
                "title": "Buy groceries",
              },
              {
                "completed": true,
                "description": "",
                "id": "2",
                "title": "Do laundry",
              },
              {
                "completed": false,
                "description": "Fix leaky faucet in the bathroom",
                "id": "3",
                "title": "Call plumber",
              },
            ],
          },
        },
        "status": 200,
      },
    },
    {
      "description": "should create a Todo with success",
      "request": {
        "body": {
          "query": "
    mutation createTodo($title: String!, $description: String) {
      createTodo(input: { title: $title, description: $description }) {
        id
        title
        description
        completed
      }
    }
  ",
          "variables": {
            "title": "Buy groceries",
          },
        },
        "headers": {
          "content-type": "application/json",
        },
        "method": "POST",
        "path": "/graphql",
      },
      "response": {
        "body": {
          "data": {
            "createTodo": {
              "completed": false,
              "description": "Milk, bread, eggs, cheese",
              "id": "1",
              "title": "Buy groceries",
            },
          },
        },
        "status": 200,
      },
    },
    {
      "description": "should found a todo item by its id",
      "providerState": "there is an existing todo item with this id",
      "request": {
        "body": {
          "query": "
    query todoById($id: ID) {
      todoById(id: $id) {
        id
        title
        description
        completed
      }
    }
  ",
          "variables": {
            "id": "1",
          },
        },
        "headers": {
          "content-type": "application/json",
        },
        "method": "POST",
        "path": "/graphql",
      },
      "response": {
        "body": {
          "data": {
            "todoById": {
              "completed": false,
              "description": "Milk, bread, eggs, cheese",
              "id": "1",
              "title": "Buy groceries",
            },
          },
        },
        "status": 200,
      },
    },
    {
      "description": "should not found a todo item by its id",
      "request": {
        "body": {
          "query": "
    query todoById($id: ID) {
      todoById(id: $id) {
        id
        title
        description
        completed
      }
    }
  ",
          "variables": {
            "id": "1",
          },
        },
        "headers": {
          "content-type": "application/json",
        },
        "method": "POST",
        "path": "/graphql",
      },
      "response": {
        "body": {
          "errors": [
            {
              "message": "The todo item 1 is not found",
            },
          ],
        },
        "status": 200,
      },
    },
  ],
  "metadata": {
    "pactSpecification": {
      "version": "2.0.0",
    },
  },
  "provider": {
    "name": "graphql-provider",
  },
}
`;
